### ISSUE TITLE
BBC: Bleadperl breaks RURBAN/YAML-Safe-0.81.tar.gz

### ISSUE THREAD

**Issue by andk** [Perl/perl5#20759]
**Description**

With v5.37.6-66-gdd6f5699e3 compilation started to fail for RURBAN/YAML-Safe-0.81.tar.gz . @rurban, you will probably be interested.

Diagnostics:

``` text
In file included from /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6-66-gdd6f5699e3/af11/lib/5.37.7/x86_64-linux-ld/CORE/perl.h:7732,
                 from perl_libyaml.h:9,
                 from Safe.xs:1:
/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6-66-gdd6f5699e3/af11/lib/5.37.7/x86_64-linux-ld/CORE/inline.h: In function ‘Perl_newPADxVOP’:
/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6-66-gdd6f5699e3/af11/lib/5.37.7/x86_64-linux-ld/CORE/inline.h:3114:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
 3114 |     OP *o = newOP(type, flags);
      |     ^~
cc1: some warnings being treated as errors
make: *** [Makefile:397: Safe.o] Error 1
  RURBAN/YAML-Safe-0.81.tar.gz
  /usr/bin/make -- NOT OK
```

Bisect leads me to:

``` text
commit dd6f5699e38163de743c93d2641b8ecff455f67c
Author: Paul "LeoNerd" Evans <leonerd@leonerd.org.uk>
Date:   Tue Dec 6 23:32:48 2022 +0000

    Define a newPADxVOP() convenience function
```
All of the above is only valid for unthreaded perls. For threaded perls there was a breakage already with v5.35.5-32-g9c91314886.

For this older breakage we have this diagnostics:

``` text
In file included from /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/perl.h:7135,
                 from perl_libyaml.h:9,
                 from Safe.xs:1:
/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/inline.h: In function ‘Perl_cop_file_avn’:
/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/inline.h:3439:5: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]
 3439 |     const char *file = CopFILE(cop);
      |     ^~~~~
Safe.c: In function ‘XS_YAML__Safe_DESTROY’:
/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/perl.h:7371:19: warning: unused variable ‘my_cxtp’ [-Wunused-variable]               
 7371 |         my_cxt_t *my_cxtp = (my_cxt_t *)PL_my_cxt_list[MY_CXT_INDEX]
      |                   ^~~~~~~
/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/perl.h:7371:19: note: in definition of macro ‘dMY_CXT’
 7371 |         my_cxt_t *my_cxtp = (my_cxt_t *)PL_my_cxt_list[MY_CXT_INDEX]
      |                   ^~~~~~~
Safe.xs: In function ‘XS_YAML__Safe_new’:
/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/perl.h:7371:19: warning: unused variable ‘my_cxtp’ [-Wunused-variable]               
 7371 |         my_cxt_t *my_cxtp = (my_cxt_t *)PL_my_cxt_list[MY_CXT_INDEX]
      |                   ^~~~~~~
[...]
 7371 |         my_cxt_t *my_cxtp = (my_cxt_t *)PL_my_cxt_list[MY_CXT_INDEX]
      |                   ^~~~~~~
/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/perl.h:7371:19: note: in definition of macro ‘dMY_CXT’
 7371 |         my_cxt_t *my_cxtp = (my_cxt_t *)PL_my_cxt_list[MY_CXT_INDEX]
      |                   ^~~~~~~
Safe.c: In function ‘XS_YAML__Safe_SafeClass’:
/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/perl.h:7371:19: warning: unused variable ‘my_cxtp’ [-Wunused-variable]               
 7371 |         my_cxt_t *my_cxtp = (my_cxt_t *)PL_my_cxt_list[MY_CXT_INDEX]
      |                   ^~~~~~~
/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi/CORE/perl.h:7371:19: note: in definition of macro ‘dMY_CXT’
 7371 |         my_cxt_t *my_cxtp = (my_cxt_t *)PL_my_cxt_list[MY_CXT_INDEX]
      |                   ^~~~~~~
cc1: some warnings being treated as errors
make: *** [Makefile:396: Safe.o] Error 1
```

Bisect for the threaded perls leads me to:

``` text
commit 9c913148860b0e83e6149d37e86cdb29663ee812
Author: Tony Cook <tony@develop-help.com>
Date:   Thu Oct 21 14:16:54 2021 +1100

    Add CopFILEAVn() and use it when cleaning up COP pointers
```

**Steps to Reproduce**

`cpan -i RURBAN/YAML-Safe-0.81.tar.gz`

**Expected behavior**

Should build, test, and install the pkg

**Perl configuration**

***For the unthreaded perl***

```
# perl -V output goes here
Summary of my perl5 (revision 5 version 37 subversion 7) configuration:
  Commit id: dd6f5699e38163de743c93d2641b8ecff455f67c
  Platform:
    osname=linux
    osvers=6.0.0-5-amd64
    archname=x86_64-linux-ld
    uname='linux k93msid 6.0.0-5-amd64 #1 smp preempt_dynamic debian 6.0.10-2 (2022-12-01) x86_64 gnulinux '
    config_args='-Dprefix=/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6-66-gdd6f5699e3/af11 -Dmyhostname=k93msid -Dinstallusrbinperl=n -Uversiononly -Dusedevel -des -Ui_db -Dlibswanted=cl pthread socket inet nsl gdbm dbm malloc dl ld sun m crypt sec util c cposix posix ucb BSD gdbm_compat -Uuseithreads -Duselongdouble -DEBUGGING=-g'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=define
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O2 -g'
    cppflags='-fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='12.2.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='long double'
    nvsize=16
    Off_t='off_t'
    lseeksize=8
    alignbytes=16
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/x86_64-linux-gnu /usr/lib /usr/lib64
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/x86_64-linux-gnu/libc.so.6
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.36'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -g -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options:
    HAS_TIMES
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_DEVEL
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_LONG_DOUBLE
    USE_PERLIO
    USE_PERL_ATOF
  Built under linux
  Compiled at Jan 30 2023 04:42:08
  @INC:
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6-66-gdd6f5699e3/af11/lib/site_perl/5.37.7/x86_64-linux-ld
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6-66-gdd6f5699e3/af11/lib/site_perl/5.37.7
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6-66-gdd6f5699e3/af11/lib/5.37.7/x86_64-linux-ld
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.37.6-66-gdd6f5699e3/af11/lib/5.37.7

```
***For the threaded perl***
```
Summary of my perl5 (revision 5 version 35 subversion 6) configuration:
  Commit id: 9c913148860b0e83e6149d37e86cdb29663ee812
  Platform:
    osname=linux
    osvers=6.0.0-5-amd64
    archname=x86_64-linux-thread-multi
    uname='linux k93msid 6.0.0-5-amd64 #1 smp preempt_dynamic debian 6.0.10-2 (2022-12-01) x86_64 gnulinux '
    config_args='-Dprefix=/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c -Dmyhostname=k93msid -Dinstallusrbinperl=n -Uversiononly -Dusedevel -des -Ui_db -Dlibswanted=cl pthread socket inet nsl gdbm dbm malloc dl ld sun m crypt sec util c cposix posix ucb BSD gdbm_compat -Duseithreads -Uuselongdouble -DEBUGGING=-g'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O2 -g'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='12.2.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/x86_64-linux-gnu /usr/lib /usr/lib64
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/x86_64-linux-gnu/libc.so.6
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.36'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -g -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options:
    HAS_TIMES
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_DEVEL
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_REENTRANT_API
    USE_THREAD_SAFE_LOCALE
  Built under linux
  Compiled at Jan 30 2023 21:01:37
  @INC:
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/site_perl/5.35.6/x86_64-linux-thread-multi
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/site_perl/5.35.6
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6/x86_64-linux-thread-multi
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.5-32-g9c91314886/1e0c/lib/5.35.6

```

**Comment by tonycoz** [comment]
YAML::Safe deliberately enables `-Werror=declaration-after-statement`, but since we now require C99, and specifically allow mixed declarations and statement, the YAML::Safe build fails.

YAML::Safe needs to be updated to not set that warning (or make it conditional on perl version)

**Comment by tonycoz** [comment]
Fixed by YAML::Safe 0.82.
